src/crafting.cpp

```diff
@@ -1403,27 +1404,208 @@ void pick_recipes(const inventory &crafting_inv,
+            if( ts.use_from == cancel ) {
+                return;
+            }
+            tool_selections.push_back( ts );
+        }
+    }
+
+    crafter->assign_activity( is_long ? ACT_LONGCRAFT : ACT_CRAFT, rec->batch_time( batch_size ), rec->id );
+    crafter->activity.values.push_back( batch_size );
+    /* legacy support for lua bindings to last_batch and lastrecipe */
+    crafter->last_batch = batch_size;
+    crafter->lastrecipe = rec->ident;
+}
+
+std::list<item> craft_command::consume_components()
+{
+    std::list<item> used;
+
+    if( empty() ) {
+        debugmsg( "Warning: attempted to consume items from an empty craft_command" );
+        return used;
+    }
+
+    for( const auto &it : item_selections ) {
+        std::list<item> tmp = crafter->consume_items( it, batch_size ); // we consume the items we selected earlier
+        used.splice( used.end(), tmp );
+    }
+
+    for( const auto &it : tool_selections ) {
+        crafter->consume_tools( it, batch_size );
+    }
+
+    return used;
+}
+
+std::vector<item_selection> craft_command::check_item_components_missing( const inventory* map_inv )
+{
+    std::vector<item_selection> missing;
+
+    for( const auto &item_sel : item_selections ) {
+        itype_id type = item_sel.comp.type;
+        item_comp component = item_sel.comp;
+        long count = ( component.count > 0 ) ? component.count * batch_size : abs( component.count );
+
+        if ( item::count_by_charges( type ) && count > 0 ) {
+            switch( item_sel.use_from ) {
+                case use_from_player:
+                    if( !crafter->has_charges( type, count ) )
+                        missing.push_back( item_sel );
+                        break;
+                case use_from_map:
+                    if( !map_inv->has_charges( type, count ) )
+                        missing.push_back( item_sel );
+                        break;
+                case use_from_both:
+                    if( !( crafter->charges_of( type ) + map_inv->charges_of( type ) >= count ) )
+                        missing.push_back( item_sel );
+                        break;
+                default: break;
+            }
+        } else { // Counting by units, not charges
+            switch( item_sel.use_from ) {
+                case use_from_player:
+                    if( !crafter->has_amount( type, count ) )
+                        missing.push_back( item_sel );
+                        break;
+                case use_from_map:
+                    if( !map_inv->has_components(type, count) )
+                        missing.push_back( item_sel );
+                        break;
+                case use_from_both:
+                    if( !( crafter->amount_of( type ) + map_inv->amount_of( type ) >= count ) )
+                        missing.push_back( item_sel );
+                        break;
+                default: break;
```

Comment:
It might be better to remove the default case and add dummy cases for the other enumeration values.

This allows the compiler to issue warnings when somebody adds new enumeration values, but does not add cases for them in those switches.